oauth2_http/java/com/google/auth/oauth2/AwsRequestSigner.java

```diff
@@ -0,0 +1,362 @@
+    return getHexEncodedSha256Hash(canonicalRequest.toString().getBytes());
+  }
+
+  /** Task 2: Create a string to sign for Signature Version 4. */
+  private String createStringToSign(
+      String canonicalRequestHash, String xAmzDate, String credentialScope) {
+    return HASHING_ALGORITHM
+        + "\n"
+        + xAmzDate
+        + "\n"
+        + credentialScope
+        + "\n"
+        + canonicalRequestHash;
+  }
+
+  /**
+   * Task 3: Calculate the signature for AWS Signature Version 4.
+   *
+   * @param date the date used in the hashing process in YYYYMMDD format
+   */
+  private String calculateAwsV4Signature(
+      String serviceName, String secret, String date, String region, String stringToSign) {
+    byte[] kDate = sign(("AWS4" + secret).getBytes(UTF_8), date.getBytes(UTF_8));
+    byte[] kRegion = sign(kDate, region.getBytes(UTF_8));
+    byte[] kService = sign(kRegion, serviceName.getBytes(UTF_8));
+    byte[] kSigning = sign(kService, AWS_REQUEST_TYPE.getBytes(UTF_8));
+    return BaseEncoding.base16().lowerCase().encode(sign(kSigning, stringToSign.getBytes(UTF_8)));
+  }
+
+  private Map<String, String> getCanonicalHeaders(String date) {
+    Map<String, String> headers = new HashMap<>();
+    headers.put("host", uri.getHost());
+
+    // Only add the date if it hasn't been specified through the "date" header.
+    if (!additionalHeaders.containsKey("date")) {
+      headers.put("x-amz-date", date);
+    }
+
+    if (awsSecurityCredentials.getToken() != null && !awsSecurityCredentials.getToken().isEmpty()) {
+      headers.put("x-amz-security-token", awsSecurityCredentials.getToken());
+    }
+
+    // Add all additional headers.
+    for (String key : additionalHeaders.keySet()) {
+      // Header keys need to be lowercase.
+      headers.put(key.toLowerCase(), additionalHeaders.get(key));
+    }
+    return headers;
+  }
+
+  private AwsDates getDates() {
+    if (additionalHeaders.containsKey("date")) {
+      return AwsDates.fromDateHeader(additionalHeaders.get("date"));
+    }
+    if (additionalHeaders.containsKey("x-amz-date")) {
+      return AwsDates.fromXAmzDate(additionalHeaders.get("x-amz-date"));
+    }
+    return AwsDates.generateXAmzDate();
+  }
+
+  private static byte[] sign(byte[] key, byte[] value) {
+    try {
+      String algorithm = "HmacSHA256";
+      Mac mac = Mac.getInstance(algorithm);
+      mac.init(new SecretKeySpec(key, algorithm));
+      return mac.doFinal(value);
+    } catch (NoSuchAlgorithmException | InvalidKeyException ex) {
+      throw new IllegalStateException("Failed to calculate the AWS V4 Signature.", ex);
+    }
+  }
+
+  private static String getHexEncodedSha256Hash(byte[] bytes) {
+    try {
+      MessageDigest digest = MessageDigest.getInstance("SHA-256");
+      return BaseEncoding.base16().lowerCase().encode(digest.digest(bytes));
+    } catch (NoSuchAlgorithmException e) {
+      throw new IllegalStateException("Failed to compute SHA-256 hash.", e);
```

Comment:
This is not an illegal state, and this is the wrong exception type to use for this problem.

How do we know SHA-256 is supported? Does the JDK promise this?

Reply:
What exception type should I use? If it's not an unchecked exception, do you want me to wrap it in a IOException and throw it in refreshAccessToken?

Good question - I'd look into it.

Reply:
NoSuchAlgorithmException is correct unless there's something wrong with the VM if SHA-256 isn't supported.

Reply:
If I throw NoSuchAlgorithmException here, I will have to handle it in refreshAccessToken in some way - I can't throw it directly. 

Also:

"Every implementation of the Java platform is required to support the following standard MessageDigest algorithms:
MD5
SHA-1
SHA-256"

https://docs.oracle.com/javase/7/docs/api/java/security/MessageDigest.html

Reply:
In that case I think a RuntimeException referencing this info is appropriate, since it means the VM isn't following spec. This is essentially a "can't happen" exception, but it's not illegal state